doc/source/ray-air/doc_code/pytorch_tabular_starter.py

```diff
@@ -0,0 +1,211 @@
+
+    # Create model.
+    model = create_model(num_features)
+    model = train.torch.prepare_model(model)
+
+    loss_fn = nn.BCELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=lr)
+
+    for cur_epoch in range(epochs):
+        for inputs, labels in to_tensor_iterator(train_data, batch_size):
+            optimizer.zero_grad()
+            predictions = model(inputs)
+            train_loss = loss_fn(predictions, labels.unsqueeze(1))
+            train_loss.backward()
+            optimizer.step()
+        validation_iterator = to_tensor_iterator(val_data, batch_size)
+        loss = validate_epoch(validation_iterator, model, loss_fn)
+        session.report({"loss": loss}, checkpoint=to_air_checkpoint(model))
+
+
+num_features = len(schema_order)
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_loop_per_worker,
+    train_loop_config={
+        # Training batch size
+        "batch_size": 128,
+        # Number of epochs to train each task for.
+        "num_epochs": 20,
+        # Number of columns of datset
+        "num_features": num_features,
+        # Optimizer args.
+        "lr": 0.001,
+    },
+    scaling_config={
+        # Number of workers to use for data parallelism.
+        "num_workers": 2,
+        # Whether to use GPU acceleration.
+        "use_gpu": False,
+        # trainer_resources=0 so that the example works on Colab.
+        "trainer_resources": {"CPU": 0},
+    },
+    datasets={"train": train_dataset, "validate": valid_dataset},
+    preprocessor=preprocessor,
+)
+
+result = trainer.fit()
+print(f"Last result: {result.metrics}")
+# Last result: {'loss': 0.6559339960416158, ...}
+# __air_pytorch_train_end__
+
+# __air_pytorch_tuner_start__
+from ray import tune
+from ray.tune.tuner import Tuner, TuneConfig
+from ray.air.config import RunConfig
+
+tuner = Tuner(
+    trainer,
+    param_space={"train_loop_config": {"lr": tune.uniform(0.0001, 0.01)}},
+    tune_config=TuneConfig(num_samples=5, metric="loss", mode="min"),
+)
+result_grid = tuner.fit()
+best_result = result_grid.get_best_result()
+print("Best Result:", best_result)
+# Best Result: Result(metrics={'loss': 0.278409322102863, ...})
+
+checkpoint = best_result.checkpoint
+# __air_pytorch_tuner_end__
+
+# __air_pytorch_batchpred_start__
+from ray.train.batch_predictor import BatchPredictor
+from ray.train.torch import TorchPredictor
+
+batch_predictor = BatchPredictor.from_checkpoint(
+    checkpoint, TorchPredictor, model=create_model(num_features)
+)
+
+predicted_probabilities = batch_predictor.predict(test_dataset)
+print("PREDICTED PROBABILITIES")
+predicted_probabilities.show()
+# {'predictions': array([1.], dtype=float32)}
+# {'predictions': array([0.], dtype=float32)}
+# __air_pytorch_batchpred_end__
+
+# __air_pytorch_online_predict_start__
+from ray import serve
+from ray.serve.model_wrappers import ModelWrapperDeployment
+
+from fastapi import Request
+import requests
+
+
+async def adapter(request: Request):
```

Comment:
Would you be able to use the built in pandas adapter instead of implementing your own?

Reply:
going to remove this, but we should simplify this also for the rest of the examples (I copied this from somewhere else)